Revamp the way OIDs are dispatched to segments on CREATE statements.

hlinnaka opened the pull request:
This PR changes the way OIDs are dispatched from master to the QE nodes. Quoting new comments in oid-dispatch.c (which is a whole new file): 

```
 * In Greenplum, it's important that most objects, like relations,
 * functions, operators, have the same OIDs in the master and all QE nodes.
 * Otherwise query plans generated in the master will not work on the
 * QE nodes, because they use the master's OIDs to refer to objects.
 *
 * Whenever a CREATE statement, or any other command that creates new objects,
 * is dispatched, the master also needs to tell the QE servers which OIDs to use
 * for the new objects. Before GPDB 5.0, that was done by modifiying all the
 * structs representing the DDL statements, i.e. DefineStmt, CreateOpClassStmt,
 * and so forth, by adding a new OID field to them. That was annoying when
 * merging with the upstream, because it required scattered changes to all the
 * structs, and the accompanying routines to copy and serialize them. Moreover,
 * for more complicated object types, like a table, a single OID was not enough,
 * as CREATE TABLE not only creates the entry in pg_class, it also creates a
 * composite type, an array type for the composite type, and possibly the same
 * for the associated toast table. 
 *
 * Starting with GPDB 5.0, we take a different tact. Whenever an OID is assigned
 * for a system table in the QD node, the OID is recorded in private memory, along
 * with the key for that object. See AddDispatchOidFromTuple() function, that
 * does that. For example, when a new type is created, the OID of the new type,
 * and the namespace and the type name of the new type, are recorded. This is
 * the 'dispatch_oids' list.  When the command is dispatched to the QE servers, all
 * the recorded OIDs are included in the dispatched request, and the QEs in turn
 * stash the list into backend-private memory. This is the 'preassigned_oids' list.
 *
 * In a QE node, when we are about to create a new object, and assign an OID for
 * it, we look into the 'preassigned_oids' list to see if we had received an OID
 * to use for the named object from the master. Under normal circumstances, we
 * should have a pre-assigned OID for all objects created in QEs, and the
 * GetPreassigned*() functions will throw an error if we don't.
 *
 * All in all, this provides a generic mechanism for DDL commands, to record OIDs
 * that are assigned for new objects in the master, transfer them to QE nodes when
 * the DDL command is dispatched, and for the QE nodes to use the same, pre-assigned,
 * OIDs for the objects.
 *
 * This same mechanism can be used to preserve OIDs during pg_upgrade. In PostgreSQL,
 * pg_upgrade only needs to preserve the OIDs of a few objects, like types, but in
 * GPDB we need to preserver most OIDs, because they need to be kept in sync between
 * the nodes. (Strictly speaking, we only need to ensure that all the nodes use the
 * same OIDs in the upgraded clusters, but they wouldn't need to be the same as
 * before upgrade. However, the most straightforward way to achieve that is to
 * use the same OIDs as before upgrade.)
 *
 * pg_upgrade has it's own mechanism to record the OIDs from the old cluster (XXX:
 * to be written), but when restoring the schema in the new cluster, it uses the same
 * 'preassigned_oids' list to restore them, that we use to assign specific OIDs
 * in a QE node at dispatch.
```

This implements the scheme that I outlined in the comments at https://github.com/hlinnaka/gpdb/pull/6. There are a few regression failures that will need to be investigated, and a few FIXME comments, but overall, this worked out quite nicely.

This PR is a net reduction, measured in lines of code, but even more importantly, it significantly reduces our diff vs. upstream.

After writing this, I wonder if we should push the OID assignments to even lower level. As this patch stands, all the functions that call simple_heap_insert() to create an object, have to call GetPreassignedOidForXXX() function to get the OID that the QD gave it. But we could push it into heap_insert() itself, like I did the "catching" of the OIDs in the master within heap_insert(). So heap_insert() could check if the table we're inserting is a system table, and if so, it could look up the OID to assign to it from the pre-assigned list, without any co-operation from the caller. That would further isolate the changes to oid-dispatch.c, rather than sprinkling them over all the functions like DefineRelation(), DefineType() etc.

This doesn't do anything for pg_upgrade yet. The idea is that pg_upgrade would also populate the preassigned_oids list, with SQL callable functions, like "set_next_array_pg_type_oid("my_type_name", 1234);". Or perhaps provide get a single map of all the OIDs in the old cluster, with a single call, and load it in in a single function call. But after they're loaded into preassigned_oids_list, one way or another, the actual CREATE commands will pick them up just as they do in the QE nodes, during normal operation.

@davecramer, @danielgustafsson, please have a look. If the catalog & storage team could also have a look, that'd be great.

ivannovick commented:
Looks promising improvement!!! @xiong-gang @asimrp can you have a look

hlinnaka commented:
One point worth mentioning is that I haven't paid much attention to performance. All the OIDs are stored in a linked list, and every time an OID is needed, the list is traversed to find it. That might become slow when creating very complicated objects, like a partitioned table with lots of partitions and lots of constraints in each. It'll be fairly straightforward and isolated change to replace the linked lists with hash tables, if needed.

hlinnaka commented:
@asimrp: Question: Do relfilenodes (i.e. pg_class.relfilenode, which is also the name of the physical file backing a table) need to be kept in sync across all server? I thought saw some code to sync them, so I included mechanism to dispatch that into this patch. But in some cases, it seems that they are not synchronized. One such case is when a bitmap index is vacuumed: the index is reindexed, which assigns a new relfilenode, but only on segments where there were some dead tuples.

I'm guessing that they don't need to be synchronized, so the code in this PR to do so (i.e. GetPreassignedRelfilenodeForRelation()) could be removed.

hlinnaka commented:
I tried removing the code to keep relfilenodes synchronized, and got a bunch of "oid XXX is already used" errors from installcheck-good as a result. So it seems that we can't simply drop it.

I think what's happening is that if one backend creates a new table, and another causes a new relfilenode to be assigned to an existing table at the same time, they may choose the same OID/relfilenode. When creating a new table, the master asks all the segments for their highest used OID (completely ignoring OID wraparound, but that's a separate issue), and then takes the maximum across across all nodes, plus 10 for some extra headroom. But while the master does all that, one of the segments might've independently used up the same OID for the relfilenode of another table. The master doesn't normally fall for that trap, because it also has an OID counter of its own so two backends in the master will never choose the same OID, but a backend in one of the segments can choose an OID that the master has also chosen concurrently.

I think we need to put some more thought onto OID management in general. Perhaps some kind of a two-phase protocol, where the master and the segments negotiate an OID that's not in use in any of the segments yet, and reserves it in all the segments at the same time. But this PR retains the status quo, so that relfilenodes are usually, but not always, chosen in the master.

ivannovick commented:
"I think we need to put some more thought onto OID management in general. "  That sounds like a great idea!!!  and avoid future headaches.

hlinnaka commented:
Rebased, and added a new commit that pushes the GetPreassignedXXX() calls down into heap_insert, as I alluded to earlier. I had to keep the specialized GetPreassignedXXX functions for a few more complicated cases, unfortunately. When creating a relation, for example, we need to get the OID of the corresponding composite type, before actually creating the composite type. Nevertheless, it's a significant improvement.

asimrp commented:
@hlinnaka, reilfilenodes need not be in sync on master and segments.  Segments map OIDs to relfilenodes using local copy of `pg_class`.  Your explanation of the "OID already in use" errors sounds correct to me - a segment choosing a relfilenode concurrently with master choosing an OID (for a different transaction) may cause the same number being chosen in both cases.

asimrp commented on src/backend/catalog/aoblkdir.c:
```diff
@@ -22,8 +22,7 @@
 void
-AlterTableCreateAoBlkdirTableWithOid(Oid relOid, Oid newOid, Oid newIndexOid,
-									 Oid * comptypeOid, bool is_part_child)
+AlterTableCreateAoBlkdirTableWithOid(Oid relOid, bool is_part_child)
```
While at it, can we please rename the function so that `WithOid` suffix is removed?

asimrp commented on src/backend/catalog/index.c:
```diff
@@ -1450,16 +1450,17 @@ setNewRelfilenodeToOid(Relation relation, TransactionId freezeXid, Oid newrelfil
+	if (Gp_role == GP_ROLE_EXECUTE)
+		newrelfilenode = GetPreassignedRelfilenodeForRelation(RelationGetRelid(relation));
+
 	if (newrelfilenode == InvalidOid)
```
This `if ()` should be replaced with `else` to make it more readable.

davecramer replied:
looks like they aren't related

asimrp commented on src/backend/catalog/oid-dispatch.c:
```diff
@@ -0,0 +1,630 @@
+ *
+ * This same mechanism can be used to preserve OIDs during pg_upgrade. In PostgreSQL,
+ * pg_upgrade only needs to preserve the OIDs of a few objects, like types, but in
+ * GPDB we need to preserver most OIDs, because they need to be kept in sync between
```
s/preserver/preserve/

asimrp commented on src/backend/catalog/oid-dispatch.c:
```diff
@@ -0,0 +1,630 @@
+
+	MemoryContextSwitchTo(oldcontext);
+
+	//elog(NOTICE, "adding OID assignment: catalog %u, namespace: %u, name: \"%s\": %u",
```
Please remove the comment.

hlinnaka commented:
On 31 October 2016 17:18:42 EET, Asim RP <notifications@github.com> wrote:
>asimrp commented on this pull request.
>
>
>
>> @@ -834,7 +834,7 @@ GetNewOid(Relation relation)
> 	 * RelationNeedsSynchronizedOIDs()).
> 	 */
>	if (Gp_role == GP_ROLE_EXECUTE &&
>RelationNeedsSynchronizedOIDs(relation))
>-		elog(WARNING, "allocated OID %u for relation \"%s\" in segment",
>+		elog(PANIC, "allocated OID %u for relation \"%s\" in segment",
>
>Should we PANIC before calling `GetNewOidWithIndex()`?  If we do that
>we can save an unnecessary increment to 
>`ShmemVariableCache->oidCount`.

Sorry, I didn't mean to include that WARNING->PANIC change in the patch, it was for debugging purposes only. ERROR might be appropriate though.

- Heikki

danielgustafsson commented on src/include/executor/execdesc.h:
```diff
@@ -191,6 +195,20 @@ typedef struct QueryDispatchDesc
+
+	/* Key data. */
+	Oid			catalog;		/* OID of the catalog table, e.g. pg_class */
+	char	   *objname;		/* object name (e.g. relation name) */
```
Any reason to not use `const char *` here and instead cast the `NameStr()` invocations to `const` rather than casting the consts to `char *` in `GetPreassignedOidFor..()`?

hlinnaka replied:
We don't usually use "const" in structs like this. Admittedly it's ugly to cast away the consts, but didn't see a nicer way to do it..

danielgustafsson commented on src/backend/catalog/oid-dispatch.c:
```diff
@@ -0,0 +1,641 @@
+ * Starting with GPDB 5.0, we take a different tact. Whenever an OID is assigned
+ * for a system table in the QD node, the OID is recorded in private memory, in
+ * the 'dispatch_oids' list, along with the key for that object.
+ * AddDispatchOidFromTuple() function does that, and it's called heap_insert().
```
_it's called_ **in** _heap_insert()._

danielgustafsson commented on src/backend/catalog/oid-dispatch.c:
```diff
@@ -0,0 +1,641 @@
+		OidAssignment *p = (OidAssignment *) lfirst(cur_item);
+
+		if (searchkey->catalog == p->catalog &&
+			(searchkey->objname == NULL || strcmp(searchkey->objname, p->objname) == 0) &&
```
`p->objname` is NULL in the case of a relfilenode from `AddDispatchRelfilenodeForRelation()` if I'm reading this right, won't we need something along the lines of below to guard against this?
```diff
- (searchkey->objname == NULL || strcmp(searchkey->objname, p->objname) == 0) &&
+ (searchkey->objname == NULL || (p->objname != NULL && strcmp(searchkey->objname, p->objname) == 0)) &&
```

xiong-gang commented on src/backend/commands/functioncmds.c:
```diff
@@ -1994,9 +1993,6 @@ CreateCast(CreateCastStmt *stmt)
 
-	if (stmt->castOid != 0)
-		HeapTupleSetOid(tuple, stmt->castOid);
-
```
Looks like the previous commit of removing caql delete one line here:
stmt->castOid = caql_insert(pcqCtx, tuple);

It leads to the inconsistent of the pg_cast oid on master and segment, which means we need to stash castOid as well.

danielgustafsson commented on src/backend/catalog/index.c:
```diff
@@ -1423,11 +1427,7 @@ index_update_stats(Relation rel, bool hasindex, bool isprimary,
 Oid
 setNewRelfilenode(Relation relation, TransactionId freezeXid)
```
The comment is out of date referring to `newrelfilenode`.

ashwinstar commented on src/include/executor/execdesc.h:
```diff
@@ -191,6 +195,20 @@ typedef struct QueryDispatchDesc
+	Oid			catalog;		/* OID of the catalog table, e.g. pg_class */
+	char	   *objname;		/* object name (e.g. relation name) */
+	Oid			namespaceOid;	/* namespace OID for most objects */
+	Oid			keyOid2;			/* */
```
What's this used for ?

danielgustafsson replied:
If I read it right `keyOid2` is used as a "misc" member in the struct to store data specific to the current catalog without requiring OidAssignment to have a specialized member for each potential use case. I agree that `keyOid2` is a bit obfuscated though, perhaps we should rename it to reflect its use or use a `union` on it?

ashwinstar commented on src/backend/catalog/oid-dispatch.c:
```diff
@@ -0,0 +1,641 @@
+			*exempt = true;
+			 break;
+
+		default:
```
Is DEFAULT case really needed ? or can we have assert or error to easily catch missing to add case here for future or new catalog tables.

hlinnaka replied:
As the patch stands, the caller does throws an ERROR or WARNING, if the catalog table is not recognized. That gives a slightly better error message, although this is a "shouldn't happen" case, so throwing an ERROR directly from here would be OK too. I left it as it is for now. It can be useful, for debugging, to distinguish between the case that a QE needs an OID for an unrecognized catalog table, and the case that the QD tries assigns an OID for an unrecognized catalog table.

ashwinstar commented on src/backend/catalog/oid-dispatch.c:
```diff
@@ -0,0 +1,641 @@
+ * If found, the entry is removed from the list, and the OID is returned.
+ */
+static Oid
+FindPreassignedOid(OidAssignment *searchkey)
```
Frankly, its confusing to see Find* and Get* interface internally removing the node. The function name maybe someway should convey the same or not say Find* or Get*.

ashwinstar commented:
On Fri, Oct 28, 2016 at 2:14 PM, Heikki Linnakangas <
notifications@github.com> wrote:

> One point worth mentioning is that I haven't paid much attention to
> performance. All the OIDs are stored in a linked list, and every time an
> OID is needed, the list is traversed to find it. That might become slow
> when creating very complicated objects, like a partitioned table with lots
> of partitions and lots of constraints in each. It'll be fairly
> straightforward and isolated change to replace the linked lists with hash
> tables, if needed.
>
Yes feel like linked-list would slow-down the things in case mentioned for
partitioned tables and more so if they are AO/CO with constraints and all.
Would be great to find the numbers for something like
```
CREATE TABLE sales (
        trans_id int,
        date date,
        amount decimal(9,2),
        region text)
DISTRIBUTED BY (trans_id)
PARTITION BY RANGE (date)
        SUBPARTITION BY LIST (region)
        SUBPARTITION TEMPLATE (
                SUBPARTITION usa VALUES ('usa'),
                SUBPARTITION asia VALUES ('asia'),
                SUBPARTITION europe VALUES ('europe'),
                DEFAULT SUBPARTITION other_regions)
        (
                START (date '2011-01-01') INCLUSIVE
                END (date '2012-01-01') EXCLUSIVE
                EVERY (INTERVAL '1 day'),
                DEFAULT PARTITION outlying_dates );
```
and also if instead made it a AO table.

hlinnaka commented:
On 11/03/2016 01:24 AM, Ashwin Agrawal wrote:
> On Fri, Oct 28, 2016 at 2:14 PM, Heikki Linnakangas <
> notifications@github.com> wrote:
>
>> One point worth mentioning is that I haven't paid much attention to
>> performance. All the OIDs are stored in a linked list, and every time an
>> OID is needed, the list is traversed to find it. That might become slow
>> when creating very complicated objects, like a partitioned table with lots
>> of partitions and lots of constraints in each. It'll be fairly
>> straightforward and isolated change to replace the linked lists with hash
>> tables, if needed.
>>
> Yes feel like linked-list would slow-down the things in case mentioned for
> partitioned tables and more so if they are AO/CO with constraints and all.
> Would be great to find the numbers for something like
> ```
> CREATE TABLE sales (
>         trans_id int,
>         date date,
>         amount decimal(9,2),
>         region text)
> DISTRIBUTED BY (trans_id)
> PARTITION BY RANGE (date)
>         SUBPARTITION BY LIST (region)
>         SUBPARTITION TEMPLATE (
>                 SUBPARTITION usa VALUES ('usa'),
>                 SUBPARTITION asia VALUES ('asia'),
>                 SUBPARTITION europe VALUES ('europe'),
>                 DEFAULT SUBPARTITION other_regions)
>         (
>                 START (date '2011-01-01') INCLUSIVE
>                 END (date '2012-01-01') EXCLUSIVE
>                 EVERY (INTERVAL '1 day'),
>                 DEFAULT PARTITION outlying_dates );
> ```
> and also if instead made it a AO table.

I tested that, and it turns out that it's OK. In CREATE TABLE ... 
PARTITION BY, We actually dispatch the creation of each partition 
separately, so the list of OIDs in each dispatched command is small. 7-8 
OIDs in each dispatch with the above command.

It would be nice to not dispatch them separately, as that adds a lot of 
round-trips, and is probably the biggest reason why creating a table 
with a lot of partitions is so slow. But until we do, a linked list will 
probably perform OK.

- Heikki

ivannovick commented:
"It would be nice to not dispatch them separately, as that adds a lot of
round-trips, and is probably the biggest reason why creating a table
with a lot of partitions is so slow" any idea how to resolve this.  Sounds like a nice win if possible.

danielgustafsson commented:
Since @hlinnaka is at a conference this week I picked up this branch trying to get a few more tests green. The commits I pushed are all intended to be rebased into the original commits (or reverted in case of bogus), thus commit messages are less than perfect.

hlinnaka commented:
Thanks Daniel! I renamed oid-dispatch.[ch] to oid_dispatch.[ch], per @xiong-gang's suggestion, and fixed the things @asimrp pointed out. I rebased, included Daniel's fixes, squashed it all together, and force-pushed. Daniel, I hope I didn't joggle your elbow if you were just hacking on this.

Any more comments?

hlinnaka commented:
On 11/03/2016 03:41 PM, Ivan Novick wrote:
> "It would be nice to not dispatch them separately, as that adds a lot of
> round-trips, and is probably the biggest reason why creating a table
> with a lot of partitions is so slow" any idea how to resolve this.  Sounds like a nice win if possible.

It ought to be possible to dispatch the original CREATE TABLE command to 
all segments, and have them figure out and create the partition layout 
independently. Assuming the process is deterministic, they should all 
create the same layout.

Alternatively, instead of dispatching the commands one by one as they're 
created, collect them in a list, and submit them all in one batch after 
the master has created all the objects.

- Heikki

hlinnaka commented:
On 11/04/2016 04:46 PM, Daniel Gustafsson wrote:
>> @@ -191,6 +195,20 @@ typedef struct QueryDispatchDesc
>  	List	   *cursorPositions;
>  } QueryDispatchDesc;
>
> +typedef struct
> +{
> +	NodeTag		type;
> +
> +	/* Key data. */
> +	Oid			catalog;		/* OID of the catalog table, e.g. pg_class */
> +	char	   *objname;		/* object name (e.g. relation name) */
> +	Oid			namespaceOid;	/* namespace OID for most objects */
> +	Oid			keyOid2;			/* */
>
> If I read it right `keyOid2` is used as a "misc" member in the struct
> to store data specific to the current catalog without requiring
> OidAssignment to have a specialized member for each potential use
> case.

Yep

> I agree that `keyOid2` is a bit obfuscated though, perhaps we
> should rename it to reflect its use or use a `union` on it?

Yeah, I thought about that. Union might be good. We don't usually use 
unions in structs that we use with the readfuncs/outfuncs/etc. 
machinery, though. No fundamental reason that wouldn't work, we just 
usually don't. One option would be to have regular fields for all of the 
things, and just leave others unused when they're not needed. 
namespaceOid is also horribly abused to store other things, perhaps we 
should clean that up too.

- Heikki

danielgustafsson commented:
> On 04 Nov 2016, at 16:57, Heikki Linnakangas <notifications@github.com> wrote:
> 
> Yeah, I thought about that. Union might be good. We don't usually use 
> unions in structs that we use with the readfuncs/outfuncs/etc. 
> machinery, though. No fundamental reason that wouldn't work, we just 
> usually don’t.

To be fair, unions can be considered an "obfuscation trick” as well since they
wont be expected for just this reason.

> One option would be to have regular fields for all of the 
> things, and just leave others unused when they're not needed. 
> namespaceOid is also horribly abused to store other things, perhaps we 
> should clean that up too.

I assume that the only reason for not doing regular fields would be performance
iff a larger struct would be penalized?

asimrp commented:
Hey @hlinnaka, Ashwin and I discussed the idea of including relfilenodes in dispatch and we think it should be suffice to dispatch only OIDs.  We cannot explain the "OID already in use" errors you observed when only OIDs were included in dispatch_list.  ICG doesn't fail with "OID already in use" error on master, where only OIDs are dispatched.  The question therefore is why is this patch causing such errors.  Do you happen to have stack trace for the error?

hlinnaka commented:
On 11/04/2016 08:35 PM, Asim RP wrote:
> Hey @hlinnaka, Ashwin and I discussed the idea of including
> relfilenodes in dispatch and we think it should be suffice to
> dispatch only OIDs.  We cannot explain the "OID already in use"
> errors you observed when only OIDs were included in dispatch_list.
> ICG doesn't fail with "OID already in use" error on master, where
> only OIDs are dispatched.

We do dispatch relfilenodes in master. For some operations. 
IndexInfoOpaque has fields heapRelfilenode and indexRelfilenode for 
that. And VacuumStmt and IndexVacuumInfo have a extra_oids fields for 
that. I think they're related.

>  The question therefore is why is this
> patch causing such errors.  Do you happen to have stack trace for the
> error?

I don't have one at hand. It should be easy to reproduce. if you just 
comment out the calls to AddDispatchRelfilenodeForRelation() from this PR.

- Heikki

pivotal-issuemaster commented:
@hlinnaka Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=1263)!

[Click here](https://cla.pivotal.io/sync/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=1263) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/faq) for frequently asked questions.

pivotal-issuemaster commented:
@hlinnaka Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=1263)!
